main/src/library/Channel/Channel.flix

```diff
@@ -0,0 +1,582 @@
+/*
+ * Copyright 2019 Magnus Madsen
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+// TODO : remove this
+def printf[a, e](toString: a -> String & e, str: String, val: a): Unit & Impure =
+    let stringBuilder = StringBuilder.new();
+    let convertedStr = toString(val);
+    StringBuilder.appendString!(stringBuilder, str);
+    StringBuilder.appendString!(stringBuilder, convertedStr);
+    let fullStr = StringBuilder.toString(stringBuilder);
+    Console.printLine(fullStr)
+
+
+namespace Channel {
+
+    // todo: remove this
+    def exit(status: Int32): () & Impure =
+        import java.lang.Runtime:getRuntime();
+        let runtime = getRuntime();
+        import java.lang.Runtime.exit(Int32);
+        runtime.exit(status)
+
+    /**
+    * The tuple corresponds to (id, channelLock, bufferSize, unbuffered, elementQueue, waitingGetters, waitingSetters)
+    *   - id
+    *   - channelLock is the channelLock of this channel.
+    *   - bufferSize is the size of a channel. If you try to put an
+    *   element in a channel that's full, you wait until there's space.
+    *   - a flag for whether the channel is unbuffered or not
+    *   - elementQueue is the elementQueue of elements in the list.
+    *   - waitingGetters is a set of conditions that is waiting for get.
+    *   This set is cleared after each new element.
+    *   - waitingSetters is a condition that can notify threads of
+    *   available space in the elementQueue.
+    */
+    pub opaque type Channel[a] = (
```

Comment:
Rename this type to `ChannelImpl` or something else because otherwise there is a clash between the built-in type `Channel` and this type. Flix does not report this clash, because otherwise you _will_ get strange errors. Here be dragons :)